dali/python/nvidia/dali/_utils/dali_trace.py

```diff
@@ -0,0 +1,184 @@
+    """
+    if prev_frame_summary is None or new_frame_summary is None:
+        return False
+    return (
+        prev_frame_summary.name == new_frame_summary.name
+        and prev_frame_summary.filename == new_frame_summary.filename
+    )
+
+
+def _filter_autograph_frames(stack_summary, frame_map, frame_filter):
+    origin_stack_summary = []
+    is_ag_function_call_start = False
+    current_function_region = None
+
+    for frame_entry in stack_summary:
+        # The frame_map maps filename:lineno pairs of AG-produced code back to FrameSummary of
+        # original code.
+        ag_entry = (frame_entry.filename, frame_entry.lineno)
+
+        if _remap_ag_frames and ag_entry in frame_map:
+            origin_info = frame_map[ag_entry]
+            origin_frame_entry = traceback.FrameSummary(
+                filename=origin_info[0],
+                lineno=origin_info[1],
+                name=origin_info[2],
+                line=origin_info[3],
+            )
+        else:
+            origin_frame_entry = frame_entry
+
+        # We will skip code that is related to internal AutoGraph and conditionals implementation,
+        # leaving us with code produced by transforming user-code
+        # See api.py:converted_call for details on filtering.
+        skip = frame_filter.is_filtered(origin_frame_entry.filename)
+
+        # Detect repeated appearance of function transformed by AG
+        # AutoGraph is wrapping a function call - entry point
+        if is_frame_converted_call(frame_entry):
+            is_ag_function_call_start = True
+            current_function_region = None
+        # It quits to a non-AG code, treat it as normal from now-on
+        if is_frame_call_unconverted(frame_entry):
+            is_ag_function_call_start = False
+            current_function_region = None
+        # We are in the first part of the converted_func call that is not skipped
+        # (as we are in user code, remember the function)
+        if is_ag_function_call_start and not skip:
+            is_ag_function_call_start = False
+            current_function_region = origin_frame_entry
+            skip = True
+            origin_stack_summary.append(origin_frame_entry)
+
+        # User code - not filtered out
+        if not skip:
+            # If we are in the same function region, we replace previous entry so we keep only the
+            # last one
+            assert origin_stack_summary
+            if _is_matching_function(origin_stack_summary[-1], current_function_region):
+                if _collapse_ag_frames:
+                    origin_stack_summary.pop()
+            else:
+                current_function_region = None
+            origin_stack_summary.append(origin_frame_entry)
+        elif not _filter_ag_frames and (
+            len(origin_stack_summary) == 0 or origin_stack_summary[-1] != origin_frame_entry
+        ):
+            origin_stack_summary.append(origin_frame_entry)
```

Comment:
```suggestion
        # User code - not filtered out
        if not skip:
            if is_ag_function_call_start:
                is_ag_function_call_start = False
                current_function_region = origin_frame_entry
                origin_stack_summary.append(origin_frame_entry)
            else:    
                # If we are in the same function region, we replace previous entry so we keep only the
                # last one
                assert origin_stack_summary
                if _is_matching_function(origin_stack_summary[-1], current_function_region):
                    if _collapse_ag_frames:
                        origin_stack_summary.pop()
                else:
                    current_function_region = None
                origin_stack_summary.append(origin_frame_entry)
        elif not _filter_ag_frames:
            origin_stack_summary.append(origin_frame_entry)
```

Sorry for so intrusive suggestions, but the `skip=True` part got me confused a little. I'd rephrase it a little, first for clarity of what the skip means, second, so that you don't need to double check (in `elif not _filter_ag_frames`) if you don't add the same frame twice). The latter part could potentially filter too much (if there existed recursive internal ag calls:)), right?

Reply:
I'm up for those kind of suggestions that can make it clearer.